File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root or !p or ! q or p == root or q == root or (root->val > p->val and root->val < q->val) or (root->val < p->val and root->val > q->val))  {
            return root;
        }
        
        if (root->val > p->val and root->val > q->val) {
            return lowestCommonAncestor(root->left, p, q);
        } else if (root->val < p->val and root->val < q->val) {
            return lowestCommonAncestor(root->right, p, q);
        }
        return NULL;
    }
};

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // lowest common ancestor - first common ancestor when traversing bottom up i.e. from the leaf nodes (if the leaf nodes are the given inputs)
        // a node can be a descendant of itself.
        // both the given nodes can be in the left sub tree or can be in the right sub tree.
        // If one is in the left and one is in the right, it is obvious that root is the lowest and only common ancestor.
        int rootVal = root.val;
        int pVal = p.val;
        int qVal = q.val;

        if (pVal > rootVal && qVal > rootVal) {
            return lowestCommonAncestor(root.right, p, q);
        } else if (pVal < rootVal && qVal < rootVal) {
            return lowestCommonAncestor(root.left, p, q);
        } else {
            return root;
        }
    }
}